extensions/amp-auto-ads/0.1/ad-strategy.js

```diff
@@ -62,14 +88,23 @@ export class AdStrategy {
       dev().warn(TAG, 'unable to fulfill ad strategy');
       return Promise.resolve(false);
     }
-    return nextPlacement.placeAd(this.type_, this.baseDataAttributes_)
-        .then(state => {
-          if (state == PlacementState.PLACED) {
-            this.placedAds_.push(nextPlacement);
-            return true;
-          } else {
-            return this.placeNextAd_();
-          }
-        });
+    return nextPlacement.getEstimatedPosition().then(yPosition => {
+      return this.adTracker_.isTooNearAnAd(
+          yPosition, this.strategyConfig_.minAdSpacing).then(tooNear => {
+            if (tooNear) {
```

Comment:
This duplication (it's minimal, but there) could be reduced if we move this "too near" detection into `Placement#placeAd`.

```js
nextPlacement.placeAd(...).then(state => {
  if (state == PLACED) {
  } else {
    // too near, or some other failure
  }
});